File: Difficulty: Medium/Job Sequencing Problem/job-sequencing-problem.js
//{ Driver Code Starts
// Initial Template for javascript
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => { inputString += inputStdin; });

process.stdin.on('end', _ => {
    inputString =
        inputString.trim().split('\n').map(string => { return string.trim(); });

    main();
});

function readLine() { return inputString[currentLine++]; }

function main() {
    let t = parseInt(readLine());
    while (t--) {
        let jobIDs = readLine().split(' ').map(x => parseInt(x));
        let deadlines = readLine().split(' ').map(x => parseInt(x));
        let profits = readLine().split(' ').map(x => parseInt(x));
        let obj = new Solution();
        let ans = obj.JobSequencing(jobIDs, deadlines, profits);
        console.log(ans[0] + " " + ans[1]);
    }
}
// } Driver Code Ends

// User-defined function Template for JavaScript  

/**  
 * Schedules jobs to maximize profit while meeting deadlines.  
 * @param {number[]} jobIds - Array of job IDs  
 * @param {number[]} deadlines - Array of job deadlines  
 * @param {number[]} profits - Array of job profits  
 * @param {number} jobCount - Total number of jobs  
 * @returns {number[]} - An array containing the number of jobs scheduled and the maximum profit  
 */  

class Solution {  
    // Function to schedule jobs based on maximum profit and deadline constraints  
    JobSequencing(jobIds, deadlines, profits) {  
        // Array to store job details (ID, deadline, profit)  
        let jobList = [];  

        for (let i = 0; i < deadlines.length; i++) {  
            jobList.push([jobIds[i], deadlines[i], profits[i]]);  
        }  

        // Sort jobs in descending order of profit  
        jobList.sort((a, b) => b[2] - a[2]);  

        // Determine the maximum deadline value  
        let maxDeadline = 0;  
        for (let d of deadlines) {  
            maxDeadline = Math.max(d, maxDeadline);  
        }  

        // Array to track scheduled jobs, initialized with -1 (empty slots)  
        let schedule = new Array(maxDeadline + 1).fill(-1);  

        let totalProfit = 0;  
        let jobCount = 0;  

        // Iterate through sorted jobs to schedule them  
        for (let job of jobList) {  
            let [jobId, deadline, profit] = job;  

            // Find the latest available slot before or on the deadline  
            for (let i = deadline; i > 0; --i) {  
                if (schedule[i] == -1) {  
                    schedule[i] = jobId;  
                    totalProfit += profit;  
                    ++jobCount;  
                    break;  
                }  
            }  
        }  

        // Return the number of jobs scheduled and the maximum profit  
        return [jobCount, totalProfit];  
    }  
}  

File: Difficulty: Medium/Job Sequencing Problem/job-sequencing-problem.cpp
//{ Driver Code Starts
// Program to find the maximum profit job sequence from a given array 
// of jobs with deadlines and profits 
#include<bits/stdc++.h>
using namespace std; 

// A structure to represent a job 
struct Job 
{ 
    int id;	 // Job Id 
    int dead; // Deadline of job 
    int profit; // Profit if job is over before or on deadline 
}; 


// } Driver Code Ends


/* 
   Structure to represent a Job
   Each job has an ID, a deadline, and a profit if the job is completed before or on the deadline.

struct Job 
{ 
    int id;    // Job ID 
    int dead;  // Deadline of the job 
    int profit; // Profit if the job is completed before or on its deadline 
};

*/

class Solution 
{
    public:
    // Comparator function to sort jobs in descending order of profit
    static int compareProfit(Job a, Job b) {
        return a.profit > b.profit; // Return true if profit of a is greater than b
    }
    
    // Function to schedule jobs in such a way that the profit is maximized
    vector<int> JobScheduling(Job jobs[], int n) 
    { 
        int totalProfit = 0;  // To store the total profit from the scheduled jobs
        int totalJobs = 0;    // To store the count of jobs that are scheduled

        // Sort jobs in descending order of profit using the comparator
        sort(jobs, jobs + n, compareProfit);

        int jobResult[n];     // Array to store the job ID assigned to each slot
        int slots[n] = {0};   // Array to track if a time slot is occupied (1 = occupied, 0 = free)

        // Iterate through each job to assign it to an available time slot
        for(int i = 0; i < n; i++) {
            // Try to find the latest available slot before the job's deadline
            for(int j = jobs[i].dead - 1; j >= 0; j--) {
                if(slots[j] == 0) { // If slot is free
                    jobResult[j] = i; // Assign the job to the current slot
                    slots[j] = 1;      // Mark the slot as occupied
                    break;             // Move to the next job
                }
            }
        }

        // Calculate the total number of jobs scheduled and the total profit
        for(int i = 0; i < n; i++) {
            if(slots[i]) {  // If the slot is occupied
                totalJobs++;               // Increment the count of jobs scheduled
                totalProfit += jobs[jobResult[i]].profit; // Add the profit from the scheduled job
            }
        }

        // Return the total number of jobs scheduled and the total profit
        vector<int> result;
        result.push_back(totalJobs);  // Number of jobs scheduled
        result.push_back(totalProfit); // Total profit earned from scheduled jobs

        return result;
    } 
};


//{ Driver Code Starts.
// Driver program to test methods 
int main() 
{ 
    int t;
    //testcases
    cin >> t;
    
    while(t--){
        int n;
        
        //size of array
        cin >> n;
        Job arr[n];
        
        //adding id, deadline, profit
        for(int i = 0;i<n;i++){
                int x, y, z;
                cin >> x >> y >> z;
                arr[i].id = x;
                arr[i].dead = y;
                arr[i].profit = z;
        }
        Solution ob;
        //function call
        vector<int> ans = ob.JobScheduling(arr, n);
        cout<<ans[0]<<" "<<ans[1]<<endl;
    
cout << "~" << "\n";
}
	return 0; 
}



// } Driver Code Ends